finality-aleph/src/testing/proposal.rs

```diff
@@ -0,0 +1,270 @@
+type TestAuxChainInfo = AuxFinalizationChainInfoProvider<Block, Arc<TestClient>>;
+
+fn prepare_proposal_test() -> (ChainBuilder, TestCachedChainInfo, TestAuxChainInfo) {
+    let client = Arc::new(TestClientBuilder::new().build());
+
+    let config = ChainInfoCacheConfig {
+        block_cache_capacity: 2,
+    };
+    let cached_chain_info_provider = CachedChainInfoProvider::new(client.clone(), config);
+
+    let chain_builder =
+        ChainBuilder::new(client.clone(), Arc::new(TestClientBuilder::new().build()));
+
+    let aux_chain_info_provider =
+        AuxFinalizationChainInfoProvider::new(client, chain_builder.genesis_hash_num());
+
+    (
+        chain_builder,
+        cached_chain_info_provider,
+        aux_chain_info_provider,
+    )
+}
+
+fn verify_proposal_status(
+    cached_cip: &mut TestCachedChainInfo,
+    aux_cip: &mut TestAuxChainInfo,
+    proposal: &AlephProposal<Block>,
+    correct_status: ProposalStatus<Block>,
+) {
+    let status_c = get_proposal_status(cached_cip, proposal, None);
+    assert_eq!(
+        status_c, correct_status,
+        "Cached chain info gives wrong status for proposal {:?}",
+        proposal
+    );
+    let status_a = get_proposal_status(aux_cip, proposal, None);
+    assert_eq!(
+        status_a, correct_status,
+        "Aux chain info gives wrong status for proposal {:?}",
+        proposal
+    );
+}
+
+#[tokio::test]
+async fn correct_proposals_are_finalized_even_with_forks() {
+    let (mut chain_builder, mut cached_cip, mut aux_cip) = prepare_proposal_test();
+
+    let blocks = chain_builder
+        .build_and_import_branch_upon(
+            &chain_builder.genesis_hash(),
+            MAX_DATA_BRANCH_LEN * 10,
+            false,
+        )
+        .await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..len].to_vec();
+        let proposal = proposal_from_blocks(blocks_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            ProposalStatus::Finalize(proposal.top_block()),
+        );
+    }
+
+    let _forks = chain_builder
+        .build_and_import_branch_upon(&blocks[2].header.hash(), MAX_DATA_BRANCH_LEN * 10, false)
+        .await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..len].to_vec();
+        let proposal = proposal_from_blocks(blocks_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            ProposalStatus::Finalize(proposal.top_block()),
+        );
+    }
+}
+
+#[tokio::test]
+async fn not_finalized_ancestors_handled_correctly() {
+    let (mut chain_builder, mut cached_cip, mut aux_cip) = prepare_proposal_test();
+
+    let blocks = chain_builder
+        .build_and_import_branch_upon(
+            &chain_builder.genesis_hash(),
+            MAX_DATA_BRANCH_LEN * 10,
+            false,
+        )
+        .await;
```

Comment:
I think that it is quite reasonable to put such chain initialization into either `prepare_proposal_test` or aux function

Reply:
Added two functions that do that now.

Reply:
Actually, now I see that what I did can be improved. WIP on this :)

Reply:
Ok, done